src/cmakeTaskProvider.ts

```diff
@@ -162,6 +166,89 @@ export class CMakeTaskProvider implements vscode.TaskProvider {
         }
         return undefined;
     }
+
+    public static async resolveInternalTask(task: CMakeTask): Promise<CMakeTask | undefined> {
+        const execution: any = task.execution;
+        if (!execution) {
+            const definition: CMakeTaskDefinition = <any>task.definition;
+            const scope: vscode.WorkspaceFolder | vscode.TaskScope = vscode.TaskScope.Workspace;
+            const resolvedTask: CMakeTask = new vscode.Task(definition, scope, definition.label, CMakeTaskProvider.CMakeSourceStr,
+                new vscode.CustomExecution(async (resolvedDefinition: vscode.TaskDefinition): Promise<vscode.Pseudoterminal> =>
+                    new CustomBuildTaskTerminal(resolvedDefinition.command, resolvedDefinition.targets, resolvedDefinition.preset, resolvedDefinition.options)
+                ), []);
+            return resolvedTask;
+        }
+        return undefined;
+    }
+
+    public static async findBuildTask(presetName?: string, targets?: string[]): Promise<CMakeTask | undefined> {
+        // Fetch all CMake task from `tasks.json` files as well as tasks from task provider.
+        const allTasks: vscode.Task[] = await vscode.tasks.fetchTasks({ type: CMakeTaskProvider.CMakeScriptType });
+        const tasks: (CMakeTask | undefined)[] = allTasks.map((task: any) => {
+            if (!task.definition.label || !task.group || (task.group && task.group !== vscode.TaskGroup.Build)) {
+                return undefined;
+            }
+            const definition: CMakeTaskDefinition = {
+                type: task.definition.type,
+                label: task.definition.label,
+                command: task.definition.command,
+                targets: task.definition.targets,
+                preset: task.definition.preset,
+                options: task.definition.options
+            };
+
+            const buildTask: CMakeTask = new vscode.Task(definition, vscode.TaskScope.Workspace, task.definition.label, CMakeTaskProvider.CMakeSourceStr);
+            buildTask.detail = task.detail;
+            if (task.group.isDefault) {
+                buildTask.isDefault = true;
+            }
+            return buildTask;
+        });
+
+        const buildTasks: CMakeTask[] = tasks.filter((task) => task !== undefined) as CMakeTask[];
+
+        // No CMake Task is found.
+        if (buildTasks.length === 0) {
+            return undefined;
+        }
+
+        // Find tasks with a target that matches the input preset's target or the input targets
+        let matchingTargetTasks: CMakeTask[];
+
+        if (presetName) {
+            matchingTargetTasks = buildTasks.filter(task => task.definition.preset === presetName);
+        } else {
+            matchingTargetTasks = buildTasks.filter(task => {
```

Comment:
This filter does not take into account tasks defined in tasks,json which are "generic" (have no target defined).
From my perspective it is crucial!

In my project I have following task in tasks.json which should be used as build task for every target:

```
{
    "version":"2.0.0",
    "tasks": [
        {
            "label": "CMake: Build",
            "type": "cmake",
            "command": "build",
            "group": {
                "kind": "build",
                "isDefault": true
            },
            "problemMatcher": {
                "base": "$gcc",
                "source": "ArmClang",
                "fileLocation": "absolute"
            },
            "detail": "user defined"
        }
    ]
}
```


So, if there are no tasks with targets matching exactly it should also be checked if there are some tasks with targets being "unset".
I previously used following filter definition:
```
        // Get only tasks with no targets set
        const noTargetTasks = tasks.filter(task =>
            (task.definition as CMakeTaskDefinition).targets === undefined);
```